hi-ml-cpath/src/health_cpath/utils/callbacks.py

```diff
@@ -384,69 +425,119 @@ def plot_loss_heatmap_for_slides_of_epoch(
         plt.xlabel(self.X_LABEL)
         plt.ylabel(self.Y_LABEL)
         plt.title(f"Loss values evolution for {order} slides of epoch {epoch}")
-        plt.savefig(self.get_heatmap_plot_file(epoch, order), bbox_inches="tight")
+        plt.savefig(self.get_heatmap_plot_file(epoch, order, stage), bbox_inches="tight")
 
-    @torch.no_grad()
-    def on_train_batch_start(  # type: ignore
-        self, trainer: Trainer, pl_module: BaseDeepMILModule, batch: Dict, batch_idx: int, unused: int = 0,
-    ) -> None:
-        """Caches loss values per slide at each training step in a local variable self.loss_cache."""
+    @staticmethod
+    def compute_entropy(class_probs: torch.Tensor) -> List[float]:
+        """Computes the entropy of the class probabilities.
+
+        :param class_probs: The class probabilities.
+        """
+        return (-torch.sum(class_probs * torch.log(class_probs), dim=1)).tolist()
+
+    def update_loss_cache(self, trainer: Trainer, outputs: BatchResultsType, batch: Dict, stage: ModelKey) -> None:
+        """Updates the loss cache with the loss values for the current batch."""
         if self.should_cache_loss_values(trainer.current_epoch):
-            bag_logits, bag_labels, _ = pl_module.compute_bag_labels_logits_and_attn_maps(batch)
-            if pl_module.n_classes > 1:
-                loss = pl_module.loss_fn_no_reduction(bag_logits, bag_labels.long())
-            else:
-                loss = pl_module.loss_fn_no_reduction(bag_logits.squeeze(1), bag_labels.float())
-            self.loss_cache[ResultsKey.LOSS].extend(loss.tolist())
-            self.loss_cache[ResultsKey.SLIDE_ID].extend([slides[0] for slides in batch[ResultsKey.SLIDE_ID]])
+            self.loss_cache[stage][ResultsKey.LOSS].extend(outputs[ResultsKey.LOSS_PER_SAMPLE])
+            self.loss_cache[stage][ResultsKey.SLIDE_ID].extend([slides[0] for slides in batch[ResultsKey.SLIDE_ID]])
```

Comment:
Why do the slide IDs need to be retrieved from the input `batch` instead of the `outputs`?

Reply:
Oh no particular reason, I just think of slide ids as an "input" to the model that come from the batch, but might be safer to retrieve it from outputs for consistency